cpp/src/arrow/array/concatenate.cc

```diff
@@ -75,6 +75,31 @@ struct Bitmap {
   bool AllSet() const { return data == nullptr; }
 };
 
+enum class OffsetBufferOpOutcome {
+  kOk,
+  kOffsetOverflow,
+};
+
+Status OffsetOverflowStatus() {
+  return Status::Invalid("offset overflow while concatenating arrays");
+}
+
+#define RETURN_IF_NOT_OK_OUTCOME(outcome)        \
+  switch (outcome) {                             \
+    case OffsetBufferOpOutcome::kOk:             \
+      break;                                     \
+    case OffsetBufferOpOutcome::kOffsetOverflow: \
+      return OffsetOverflowStatus();             \
+  }
```

Comment:
```suggestion
#define _RETURN_IF_OUTCOME_NOT_OK(outcome, ret)                     \
  if (ARROW_PREDICT_FALSE(outcome != OffsetBufferOpOutcome::kOk)) { \
    return ret;                                                     \
  }

#define RETURN_IF_OUTCOME_NOT_OK(outcome) _RETURN_IF_OUTCOME_NOT_OK(outcome, outcome)

#define RETURN_INVALID_IF_OUTCOME_NOT_OK(outcome) \
  _RETURN_IF_OUTCOME_NOT_OK(outcome, OffsetOverflowStatus())
```

Reply:
Maybe we can use these macros universally in this PR?

Reply:
The point of using a `switch` is to get compilation errors when a new enum case is added.